File: 1/py/part2.py
import sys

inputFile = sys.argv[1]
input = ""

with open(inputFile, "r") as f:
    input = f.read()

left_side = {}
right_side = {}
similarity_score = 0

for row in input.split("\n"):
    splitted = row.split("   ")
    first_value = int(splitted[0])
    if first_value not in left_side:
        left_side[first_value] = 0
    left_side[first_value] += 1

    second_value = int(splitted[-1])
    if second_value not in right_side:
        right_side[second_value] = 0
    right_side[second_value] += 1

for key, value in left_side.items():
    if key in right_side:
        similarity_score += key*value*right_side[key]

print(similarity_score)

File: 1/go/part2.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
)

func part2(filename string) {

	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	r := bufio.NewReader(file)

	leftSide := map[int]int{}
	rightSide := map[int]int{}
	var similarityScore int

	for {
		line, _, err := r.ReadLine()
		if err != nil {
			break
		}
		if len(line) > 1 {
			values := strings.Split(string(line), "   ")
			value1, err := strconv.Atoi(values[0])
			if err != nil {
				log.Fatal(err)
			}
			value2, err := strconv.Atoi(values[1])
			if err != nil {
				log.Fatal(err)
			}

			leftSide[value1] += 1
			rightSide[value2] += 1
		}
	}
	for id, count := range leftSide {
		similarityScore += id * count * rightSide[id]
	}

	log.Println(similarityScore)
}
